statsmodels/sandbox/mice/mice.py

```diff
@@ -0,0 +1,1167 @@
+  an additional 'analysis model' which is the central model of
+  interest (it may or may not be identical to the conditional model
+  for the same outcome variable).  MICE_model carries out multiple
+  imputation using the conditional models, and fits the analysis model
+  to a subset of these imputed data sets.  It returns the fitted model
+  results for these analysis models.  It is structured as an iterator,
+  so the analysis model results are obtained using `next` or via any
+  standard python iterator pattern.
+
+* 'MICE' takes both a 'MICE_data' object and an analysis model
+  specification.  It runs the multiple imputation, fits the analysis
+  models, and combines the results to produce a `MICEResults` object.
+  The summary method of this results object can be used to see the key
+  estimands and inferential quantities..
+
+Notes
+-----
+By default, to conserve memory 'MICE_data' saves very little
+information from one iteration to the next.  The data set passed by
+the user is copied on entry, but then is over-written each time new
+imputations are produced.  If using 'MICE_model' or 'MICE', the fitted
+analysis models and results are saved.  MICE_data includes a
+`history_func` hook that allows arbitrary information from the
+intermediate datasets to be saved for future use.
+
+References
+----------
+JL Schafer: 'Multiple Imputation: A Primer', Stat Methods Med Res,
+1999.
+
+T E Raghunathan et al.: 'A Multivariate Technique for Multiply
+Imputing Missing Values Using a Sequence of Regression Models', Survey
+Methodology, 2001.
+
+SAS Institute: 'Predictive Mean Matching Method for Monotone Missing
+Data', SAS 9.2 User's Guide, 2014.
+
+A Gelman et al.: 'Multiple Imputation with Diagnostics (mi) in R:
+Opening Windows into the Black Box', Journal of Statistical Software,
+2009.
+"""
+
+#TODO: Add reference http://biomet.oxfordjournals.org/content/86/4/948.full.pdf
+
+
+import pandas as pd
+import numpy as np
+import patsy
+import statsmodels.api as sm
+import statsmodels
+from collections import defaultdict
+
+# Can be replaced with scipy in version >= 0.14
+class _multivariate_normal(object):
+
+    def __init__(self, mean, cov):
+        self.mean = mean
+        self.cov = cov
+        self.cov_sqrt = np.linalg.cholesky(cov)
+
+    def rvs(self):
+        p = len(self.mean)
+        z = np.random.normal(size=p)
+        return self.mean + np.dot(self.cov_sqrt, z)
+
+
+
+_mice_data_example_1 = """
+    >>> imp = mice.MICE_data(data)
+    >>> imp.set_imputer('x1', formula='x2 + np.square(x2) + x3')
+    >>> for j in range(20):
+            imp.update_all()
+            imp.data.to_csv('data%02d.csv' % j)
+"""
+
+class MICE_data(object):
+
+    __doc__ = """\
+    Wrap a data set to allow missing data handling for MICE.
+
+    Parameters
+    ----------
+    data : array-like object
+        The data set.
+    perturbation_method : string
+        The default perturbation method
+    history_func : function
```

Comment:
I guess a more standard terminology would be to call this a `call_back` such as in the optimization functions